README.md
# Grav API Plugin

The **API Plugin** for [Grav](http://github.com/getgrav/grav) adds a REST API to Grav.

| IMPORTANT!!! This plugin is currently in development as is to be considered a **beta release**.  As such, use this in a production environment **at your own risk!**. More features will be added in the future.

# Installation

The API plugin is easy to install with GPM.

```
$ bin/gpm install api
```

Or clone from GitHub and put in the `user/plugins/api` folder.

# Principles

- Two URLs per resource: one for the resource collection, one for the individual resource
- Use plural forms
- Keep URLs as short as possible
- Use nouns as resource names

HTTP Method Names
- GET: read a resource or collection
- POST: create a resource
- PUT: update a resource
- DELETE: remove a resource

# HTTP Status Codes

Use meaningful HTTP Status Codes (TODO)

- 200: Success.
- 201: Created. Returned on successful creation of a new resource. Include a 'Location' header with a link to the newly-created resource.
- 400: Bad request. Data issues such as invalid JSON, etc.
- 403: Forbidden. Example: trying to create a page already existing
- 404: Not found. Resource not found on GET.
- 405: Not Allowed.
- 501: Not Implemented. Returned by default by non implemented API methods

# Things still missing

- Support filtering, sorting and pagination on collections
- Allow clients to reduce the number of fields that come back in the response
- Access control to each API method, for groups or single users
- Use http://fractal.thephpleague.com
- Allow plugins to add API methods
- Allow to limit access to a specific IP (range)
- Use OAuth2 to secure the API http://oauth2.thephpleague.com/
- JSONP
- Return the children of a page in the page request, if desired
- Allow to only retrieve some fields on API calls that are heavy on the system (e.g. `GET /api/pages/:page`)

# Authentication

In this current form the Grav API uses Basic Authentication, with the username and password currently set in Grav.
It's just a first implementation, willing to change that at least with digest authentication.

Users with `admin.super` or `admin.api` permissions can access the whole API.

## Usage or Basic Authentication on the client-side

On the client side, add the authorization header of the request as follows:

```
Authorization: Basic <Base64 encoded value>
```

the base64 value is username:password

# Standard Grav API Endpoints

## Implemented

### Version

Methods that return the API plugin version. Depending on the returned value, some methods might be available or not depending on the plugin evolution. To be used by clients when interacting with the Grav API.

### GET /api/version

Get the API plugin major version.

e.g. "0" or "1"

### GET /api/version/major

Same as `GET /api/version`

e.g. "0" or "1"

### GET /api/version/minor

Get the API plugin major and minor version.

e.g. "0.1" or "1.0"

### GET /api/version/full

Get the API plugin full version.

e.g. "0.1.3"

### Pages

#### GET /api/pages

Lists the site pages

#### GET /api/pages/:page

Get a specific page

e.g. /api/pages/blog/my-first-post
e.g. /api/pages/contact

### Users

#### GET /api/users

Lists the site users

#### GET /api/users/:user

Get a specific user

e.g. /api/users/admin
e.g. /api/users/andy

## Not yet implemented

- GET /api/plugins
- GET /api/plugins/shoppingcart
- GET /api/themes
- GET /api/themes/theme
- GET /api/data
- GET /api/logs
- GET /api/config
- GET /api/config/system
- GET /api/config/plugin/:pluginName
- GET /api/backups
- GET /api/gpm
- POST /api/gpm/update
- GET /api/grav
- POST /api/grav/self-upgrade
- GET /api/version
- GET /api/debug
- .....


CHANGELOG.md
# v0.1.0
## 10/30/2015

1. [](#new)
    * Initial Release


api.php
<?php
namespace Grav\Plugin;

use Grav\Common\Grav;
use Grav\Common\Plugin;
use Grav\Common\User\User;

class ApiPlugin extends Plugin
{
    protected $route = 'api';

    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            'onPagesInitialized' => ['onPagesInitialized', 0],
        ];
    }

    public function onPagesInitialized()
    {
        $uri = $this->grav['uri'];

        if (strpos($uri->path(), $this->config->get('plugins.api.route') . '/' . $this->route) === false) {
            return;
        }

        if (!$this->isAuthorized()) {
            header('HTTP/1.1 401 Unauthorized');
            exit();
        }

        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 1);
        $resource = $paths[0];

        if ($resource) {
            $file = __DIR__ . '/resources/' . $resource . '.php';
            if (file_exists($file)) {
                require_once $file;
                $resourceClassName = '\Grav\Plugin\Api\\' . ucfirst($resource);
                $resource = new $resourceClassName($this->grav);
                $output = $resource->execute();
                $resource->setHeaders();

                echo $output;
            } else {
                header('HTTP/1.1 404 Not Found');
            }
        }

        exit();
    }

    private function isAuthorized()
    {
        if (!isset($_SERVER['PHP_AUTH_USER']) || !isset($_SERVER['PHP_AUTH_PW'])) {
            return false;
        }
        $username = $_SERVER['PHP_AUTH_USER'];
        $password = $_SERVER['PHP_AUTH_PW'];
        $user = User::load($username);
        $isAuthenticated = $user->authenticate($password);

        if ($isAuthenticated) {
            if ($this->authorize($user, ['admin.api', 'admin.super'])) {
                return true;
            }
        }

        return false;
    }

    /**
     * Checks user authorisation to the action.
     *
     * @param  string $action
     *
     * @return bool
     */
    public function authorize($user, $action)
    {
        $action = (array)$action;

        foreach ($action as $a) {
            if ($user->authorize($a)) {
                return true;
            }
        }

        return false;
    }

}


resources/users.php
<?php
namespace Grav\Plugin\Api;
require_once 'resource.php';

use Symfony\Component\Yaml\Yaml;

/**
 * Users API
 *
 * @todo: add POST action to add a new user
 * @todo: add PUT action to edit an existing user
 * @todo: add DELETE action to delete an existing user
 */
class Users extends Resource
{
    /**
     * Get the users list
     *
     * Implements:
     *
     * - GET /api/users
     *
     * @todo:
     *
     * @return array the users list
     */
    public function getList()
    {
        $users = [];
        $files = (array) glob($this->grav['locator']->findResource("account://") . '/*.yaml');

        if ($files) foreach ($files as $file) {
            $users[] = array_merge(array('username' => basename($file, '.yaml')), Yaml::parse($file));
        }

        return $users;
    }

    /**
     * Get a single user
     *
     * Implements:
     *
     * - GET /api/users/:user
     *
     * @todo:
     *
     * @return array the single user
     */
    public function getItem()
    {
        $file = $this->grav['locator']->findResource("account://") . '/' . $this->getIdentifier() . '.yaml';
        if (!file_exists($file)) {
            $this->setErrorCode(404);
            return;
        }
        return array_merge(array('username' => basename($file, '.yaml')), Yaml::parse($file));
    }
}


resources/config.php
<?php
namespace Grav\Plugin\Api;
require_once 'resource.php';

use Grav\Common\Config\ConfigFinder;
use Symfony\Component\Yaml\Yaml;

/**
 * Config API
 */
class Config extends Resource
{
    /**
     * Get the config files list
     *
     * Implements:
     *
     * - GET /api/config
     *
     * @todo:
     *
     * @return array the config files list
     */
    public function getList()
    {

        $this->finder = new ConfigFinder;
        $locator = $this->grav['locator'];
        $this->configLookup = $locator->findResources('config://');
        $this->blueprintLookup = $locator->findResources('blueprints://config');
        $this->pluginLookup = $locator->findResources('plugins://');

        $config = $this->finder->locateConfigFiles($this->configLookup, $this->pluginLookup);
        return $config;
    }

    /**
     * Get a single config file
     *
     * Implements:
     *
     * - GET /api/config/:config
     *
     * Examples:
     *
     * - GET /api/config/system
     * - GET /api/config/system.languages
     * - GET /api/config/system/languages
     * - GET /api/config/plugins
     * - GET /api/config/plugins.admin
     * - GET /api/config/plugins/admin
     *
     * @todo:
     *
     * @return array the single config file content
     */
    public function getItem()
    {
        $config = $this->grav['config'];
        return (array)$config->get(str_replace('/', '.', $this->getIdentifier()));
    }
}

resources/version.php
<?php
namespace Grav\Plugin\Api;
require_once 'resource.php';

use Symfony\Component\Yaml\Yaml;
use RocketTheme\Toolbox\File\File;

/**
 * Version API
 */
class Version extends Resource
{
    /**
     * Get API Version
     *
     * Implements:
     *
     * - GET /api/version
     */
    public function getList()
    {
        $version = $this->getVersion();
        return strtok($version, '.');
    }

    /**
     * Implements:
     *
     * - GET /api/version/major
     * - GET /api/version/minor
     * - GET /api/version/full
     */
    public function getItem()
    {
        switch($this->getIdentifier()) {
            case 'major':
                // Get API Major Version
                $version = $this->getVersion();
                return strtok($version, '.');
            case 'minor':
                // Get API Minor Version
                $version = $this->getVersion();
                return strtok($version, '.') . '.' . strtok('.');
            case 'full':
                //Get API Full Version
                return $this->getVersion();
        }

        return;
    }

    /**
     * Fetch the plugin version from the blueprints.yaml file
     */
    private function getVersion()
    {
        $locator = $this->grav['locator'];
        $path = $locator->findResource('user://plugins/api', true);
        $fullFileName = $path . DS . 'blueprints.yaml';

        $file = File::instance($fullFileName);
        $data = Yaml::parse($file->content());

        return $data['version'];
    }
}


resources/resource.php
<?php
namespace Grav\Plugin\Api;

use Grav\Common\Grav;

class Resource
{
    /**
     * Constructor.
     */
    public function __construct(Grav $grav)
    {
        $this->grav = $grav;
    }

    /**
     * Execute the request
     *
     * @return string the output to render
     */
    public function execute()
    {
        $httpMethod = $this->getMethod();

        if ($httpMethod == 'get') {
            if ($this->getIdentifier()) {
                $return = $this->getItem();
            } else {
                $return = $this->getList();
            }
        } else {
            $method = $httpMethod . 'Item';
            $return = $this->$method();
        }

        if ($return !== '') {
            return json_encode($return);
        } else {
            return '';
        }

    }

    /**
     * Get the request method
     *
     * @return string the method name, lowercase
     */
    protected function getMethod()
    {
        return strtolower($_SERVER['REQUEST_METHOD']);
    }

    /**
     * Get the resource name
     *
     * @return string the resource name
     */
    protected function getResource()
    {
        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 1);
        $resource = $paths[0];
        return $resource;
    }

    /**
     * Get the identifier name
     *
     * @return string the resource identifier name
     */
    protected function getIdentifier()
    {
        $paths = $this->grav['uri']->paths();
        $paths = array_splice($paths, 2);

        $identifier = join('/', $paths);
        return $identifier;
    }

    /**
     * Set the response headers based on the content retrieved
     *
     * Use the following response headers:
     * - ETag: An arbitrary string for the version of a representation. Make sure to include the media type in the hash value, because that makes a different representation. (ex: ETag: "686897696a7c876b7e")
     * - Date: Date and time the response was returned (in RFC1123 format). (ex: Date: Sun, 06 Nov 1994 08:49:37 GMT)
     * - Cache-Control: The maximum number of seconds (max age) a response can be cached. However, if caching is not supported for the response, then no-cache is the value. (ex: Cache-Control: 360 or Cache-Control: no-cache)
     * - Expires: If max age is given, contains the timestamp (in RFC1123 format) for when the response expires, which is the value of Date (e.g. now) plus max age. If caching is not supported for the response, this header is not present. (ex: Expires: Sun, 06 Nov 1994 08:49:37 GMT)
     * - Last-Modified: The timestamp that the resource itself was modified last (in RFC1123 format). (ex: Last-Modified: Sun, 06 Nov 1994 08:49:37 GMT)
     *
     * @todo implement
     */
    public function setHeaders()
    {
        header('Content-type: application/json');

        // Calculate Expires Headers if set to > 0
        $expires = $this->grav['config']->get('system.pages.expires');
        if ($expires > 0) {
            $expires_date = gmdate('D, d M Y H:i:s', time() + $expires) . ' GMT';
            header('Cache-Control: max-age=' . $expires);
            header('Expires: '. $expires_date);
        }

        // TODO: Set Last-Modified
        // TODO: Set the ETag
        // TODO: Set the HTTP response code
        // TODO: Set the Date
    }

    /**
     * Set the correct error code
     *
     * @todo implement all error codes
     */
    protected function setErrorCode($code) {
        switch((int)$code) {
            case 401:
                header('HTTP/1.1 401 Unauthorized');
                break;
            case 403:
                header('HTTP/1.1 403 Forbidden');
                break;
            case 404:
                header('HTTP/1.1 404 Not Found');
                break;
            case 405:
                header('HTTP/1.1 405 Not Allowed');
                break;
            case 501:
                header('HTTP/1.1 501 Not Implemented');
                break;
            default:
                header('HTTP/1.1 ' . $code);

        }
    }

    /**
     * Get list action
     *
     * @return
     */
    public function getList()
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Get item action
     *
     * @return
     */
    public function getItem()
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Post action
     *
     * @return
     */
    public function postItem()
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Put action
     *
     * @return
     */
    public function putItem()
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Delete action
     *
     * @return
     */
    public function deleteItem()
    {
        $this->setErrorCode(501); //Not Implemented
        return;
    }

    /**
     * Prepare and return POST data.
     *
     * @param array $post
     * @return array
     */
    protected function getPost()
    {
        return json_decode(file_get_contents('php://input'));
    }


    /**
     * Builds the return message.
     *
     * Example usage from a resource action:
     *
     *   ```
     *   $this->setErrorCode(403);
     *   $message = $this->buildReturnMessage('Page already exists. Cannot create a page with the same route');
     *   return $message;
     *   ```
     *
     * @param array $post
     * @return array
     */
    protected function buildReturnMessage($message) {
        return [
            'message' => $message
        ];
    }

}


resources/pages.php
<?php
namespace Grav\Plugin\Api;
require_once 'resource.php';

use Grav\Common\Filesystem\Folder;
use Grav\Common\Page\Page;
use RocketTheme\Toolbox\File\File;
use Symfony\Component\Yaml\Yaml;

/**
 * Pages API
 */
class Pages extends Resource
{
    /**
     * Get the pages list
     *
     * Implements:
     *
     * - GET /api/pages
     *
     * @return array the pages list
     */
    public function getList()
    {
        $pagesCollection = $this->grav['pages']->all();

        $return = [];

        foreach($pagesCollection as $page) {
            $return[$page->route()] = [];
            $return[$page->route()]['title'] = $page->title();
            $return[$page->route()]['url'] = $page->url();
            $return[$page->route()]['visible'] = $page->visible();
            $return[$page->route()]['isDir'] = $page->isDir();
            $return[$page->route()]['published'] = $page->published();
        }

        return $return;
    }

    /**
     * Get a single page
     *
     * Implements:
     *
     * - GET /api/pages/:page
     *
     * @return array the single page
     */
    public function getItem()
    {
        $pages = $this->grav['pages'];
        $page = $pages->dispatch('/' . $this->getIdentifier(), false);
        return $this->buildPageStructure($page);
    }

    /**
     * Create a new page
     *
     * Implements:
     *
     * - POST /api/pages/:page
     *
     * @todo:
     *
     * @return array the single page
     */
    public function postItem()
    {
        $data = $this->getPost();
        $pages = $this->grav['pages'];

        $page = $pages->dispatch('/' . $this->getIdentifier(), false);
        if ($page !== null) {
            // Page already exists
            $this->setErrorCode(403);
            $message = $this->buildReturnMessage('Page already exists. Cannot create a page with the same route');
            return $message;
        }

        $page = $this->page($this->getIdentifier());
        $page = $this->preparePage($page, $data);

        $page->save();

        $page = $pages->dispatch('/' . $this->getIdentifier(), false);

        return $this->buildPageStructure($page);
    }

    /**
     * Updates an existing page
     *
     * Implements:
     *
     * - PUT /api/pages/:page
     *
     * @todo:
     *
     * @return array the single page
     */
    public function putItem()
    {
        $data = $this->getPost();
        $pages = $this->grav['pages'];

        $page = $pages->dispatch('/' . $this->getIdentifier(), false);
        if ($page == null) {
            // Page does not exist
            $this->setErrorCode(404);
            $message = $this->buildReturnMessage('Page does not exist.');
            return $message;
        }

        $page = $this->page($this->getIdentifier());
        $page = $this->preparePage($page, $data);
        $page->save();

        $page = $pages->dispatch('/' . $this->getIdentifier(), false);

        return $this->buildPageStructure($page);
    }

    /**
     * Deletes an existing page
     *
     * If the 'lang' param is present in the request header, it only deletes a single
     * language if there are other languages set for that page.
     *
     * Implements:
     *
     * - DELETE /api/pages/:page
     *
     * @todo:
     *
     * @return bool
     */
    public function deleteItem()
    {
        $data = $this->getPost();

        $pages = $this->grav['pages'];
        $page = $pages->dispatch('/' . $this->getIdentifier(), false);
        if ($page == null) {
            // Page does not exist
            $this->setErrorCode(404);
            $message = $this->buildReturnMessage('Page does not exist.');
            return $message;
        }

        try {
            if (isset($data->lang) && count($page->translatedLanguages()) > 1) {
                $language = trim(basename($page->extension(), 'md'), '.') ?: null;
                $filename = str_replace($language, $data->lang, $page->name());
                $path = $page->path() . DS . $filename;
                $page->filePath($path);

                $page->file()->delete();
            } else {
                Folder::delete($page->path());
            }
        } catch (\Exception $e) {
            throw new \RuntimeException('Deleting page failed on error: ' . $e->getMessage());
        }

        return '';
    }

    /**
     * Build a page structure
     *
     * @todo: add commented fields
     *
     * @return array the single page
     */
    private function buildPageStructure($page) {
        return [
            'active' => $page->active(),
            'activeChild' => $page->activeChild(),
            // 'adjacentSibling' => $page->adjacentSibling(),
            'blueprintName' => $page->blueprintName(),
            //'blueprints' => $page->blueprints(),
            'children' => $page->children(),
            'childType' => $page->childType(),
            'content' => $page->content(),
            'date' => $page->date(),
            'eTag' => $page->eTag(),
            'expires' => $page->expires(),
            'exists' => $page->exists(),
            'extension' => $page->extension(),
           // 'extra' => $page->extra(),
            'file' => $page->file(),
            'filePath' => $page->filePath(),
            'filePathClean' => $page->filePathClean(),
            'folder' => $page->folder(),
            'frontmatter' => $page->frontmatter(),
            'getRawContent' => $page->getRawContent(),
            'header' => $page->header(),
            'home' => $page->home(),
            'id' => $page->id(),
            'isDir' => $page->isDir(),
            // 'isFirst' => $page->isFirst(),
            // 'isLast' => $page->isLast(),
            'isPage' => $page->isPage(),
            'language' => $page->language(),
            'lastModified' => $page->lastModified(),
            'link' => $page->link(),
            'maxCount' => $page->maxCount(),
            'menu' => $page->menu(),
            'metadata' => $page->metadata(),
            'modified' => $page->modified(),
            'modularTwig' => $page->modularTwig(),
            'modular' => $page->modular(),
            'name' => $page->name(),
            // 'nextSibling' => $page->nextSibling(),
            'order' => $page->order(),
            'orderDir' => $page->orderDir(),
            'orderBy' => $page->orderBy(),
            'orderManual' => $page->orderManual(),
            'parent' => $page->parent(),
            'path' => $page->path(),
            'permalink' => $page->permalink(),
            // 'prevSibling' => $page->prevSibling(),
            'publishDate' => $page->publishDate(),
            'published' => $page->published(),
            'raw' => $page->raw(),
            'rawMarkdown' => $page->rawMarkdown(),
            'rawRoute' => $page->rawRoute(),
            'root' => $page->root(),
            'routable' => $page->routable(),
            'route' => $page->route(),
            'routeCanonical' => $page->routeCanonical(),
            'slug' => $page->slug(),
            'summary' => $page->summary(),
            'taxonomy' => $page->taxonomy(),
            'template' => $page->template(),
            'title' => $page->title(),
            'translatedLanguages' => $page->translatedLanguages(),
            'unpublishDate' => $page->unpublishDate(),
            'untranslatedLanguages' => $page->untranslatedLanguages(),
            'url' => $page->url(),
            'visible' => $page->visible(),
        ];
    }

    /**
     * Returns edited page.
     *
     * @param bool $route
     *
     * @return Page
     */
    private function page($route = false)
    {
        $path = $route;
        if (!isset($this->pages[$path])) {
            $this->pages[$path] = $this->getPage($path);
        }
        return $this->pages[$path];
    }

    /**
     * Returns the page creating it if it does not exist.
     *
     * @todo: Copied from Admin Plugin. Refactor to use in both
     *
     * @param $path
     *
     * @return Page
     */
    private function getPage($path)
    {
        /** @var Pages $pages */
        $pages = $this->grav['pages'];

        if ($path && $path[0] != '/') {
            $path = "/{$path}";
        }

        $page = $path ? $pages->dispatch($path, true) : $pages->root();

        if (!$page) {
            $slug = basename($path);

            if ($slug == '') {
                return null;
            }

            $ppath = str_replace('\\', '/' , dirname($path));

            // Find or create parent(s).
            $parent = $this->getPage($ppath != '/' ? $ppath : '');

            // Create page.
            $page = new Page;
            $page->parent($parent);
            $page->filePath($parent->path() . '/' . $slug . '/' . $page->name());

            // Add routing information.
            $pages->addPage($page, $path);

            // Set if Modular
            $page->modularTwig($slug[0] == '_');

            // Determine page type.
            if (isset($this->session->{$page->route()})) {
                // Found the type and header from the session.
                $data = $this->session->{$page->route()};

                $header = ['title' => $data['title']];

                if (isset($data['visible'])) {
                    if ($data['visible'] == '' || $data['visible']) {
                        // if auto (ie '')
                        $children = $page->parent()->children();
                        foreach ($children as $child) {
                            if ($child->order()) {
                                // set page order
                                $page->order(1000);
                                break;
                            }
                        }
                    }

                }

                if ($data['name'] == 'modular') {
                    $header['body_classes'] = 'modular';
                }

                $name = $page->modular() ? str_replace('modular/', '', $data['name']) : $data['name'];
                $page->name($name . '.md');
                $page->header($header);
                $page->frontmatter(Yaml::dump((array)$page->header(), 10, 2, false));
            } else {
                // Find out the type by looking at the parent.
                $type = $parent->childType() ? $parent->childType() : $parent->blueprints()->get('child_type',
                    'default');
                $page->name($type . CONTENT_EXT);
                $page->header();
            }
            $page->modularTwig($slug[0] == '_');
        }

        return $page;
    }

    /**
     * Prepare a page to be stored: update its folder, name, template, header and content
     *
     * @param \Grav\Common\Page\Page $page
     * @param object                 $post
     */
    private function preparePage(\Grav\Common\Page\Page $page, $post = null)
    {
        $post = (array)$post;

        if (isset($post['order'])) {
            $order = max(0, (int) isset($post['order']) ? $post['order'] : $page->value('order'));
            $ordering = $order ? sprintf('%02d.', $order) : '';
            $slug = empty($post['folder']) ? $page->value('folder') : (string) $post['folder'];
            $page->folder($ordering . $slug);
        }

        if (isset($post['name']) && !empty($post['name'])) {
            $type = (string) strtolower($post['name']);
            $name = preg_replace('|.*/|', '', $type);
            $page->name($name);
            $page->template($type);
        }

        if (isset($post['header'])) {
            $header = $post['header'];
            $page->header((object) $header);
            $page->frontmatter(Yaml::dump((array) $page->header()));
        }

        $language = trim(basename($page->extension(), 'md'), '.') ?: null;
        $filename = str_replace($language, $post['lang'], $page->name());
        $path = $page->path() . DS . $filename;
        $page->filePath($path);

        if (isset($post['content'])) {
            $page->rawMarkdown((string) $post['content']);
            $page->content((string) $post['content']);
            $page->file()->markdown($page->rawMarkdown());
        }


        return $page;
    }
}
